crypto/multisig/src/lib.rs

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2018-2022 The MobileCoin Foundation
+
+//! Multi-signature implementations.
+
+#![no_std]
+#![deny(missing_docs)]
+
+extern crate alloc;
+
+use alloc::vec::Vec;
+use mc_crypto_digestible::Digestible;
+use mc_crypto_keys::{Ed25519SignatureError, PublicKey, Signature, Verifier};
+use prost::Message;
+use serde::{Deserialize, Serialize};
+
+/// A multi-signature: a collection of one or more signatures.
+#[derive(Clone, Deserialize, Digestible, Eq, Message, PartialEq, Serialize)]
+pub struct MultiSig<
+    S: Clone + Default + Digestible + Eq + Message + PartialEq + Serialize + Signature,
+> {
+    #[prost(message, repeated, tag = "1")]
+    signatures: Vec<S>,
+}
+
+impl<S: Clone + Default + Digestible + Eq + Message + PartialEq + Serialize + Signature>
+    MultiSig<S>
+{
+    /// Construct a new multi-signature from a collection of signatures.
+    pub fn new(signatures: Vec<S>) -> Self {
+        Self { signatures }
+    }
+}
+
+/// A set of M-out-of-N public keys.
+#[derive(Clone, Deserialize, Digestible, Eq, Message, PartialEq, Serialize)]
+#[serde(bound = "")]
+pub struct SignerSet<P: Default + PublicKey + Message> {
+    /// List of potential signers.
+    #[prost(message, repeated, tag = "1")]
+    signers: Vec<P>,
+
+    /// Minimum number of signers required.
+    #[prost(uint32, tag = "2")]
+    threshold: u32,
+}
+
+impl<P: Default + PublicKey + Message> SignerSet<P> {
+    /// Construct a new `SignerSet` from a list of public keys and threshold.
+    pub fn new(signers: Vec<P>, threshold: u32) -> Self {
+        Self { signers, threshold }
+    }
+
+    /// Verify a message against a multi-signature, returning the list of
+    /// signers that signed it.
+    pub fn verify<
+        S: Clone + Default + Digestible + Eq + Message + PartialEq + Serialize + Signature,
+    >(
+        &self,
+        message: &[u8],
+        multi_sig: &MultiSig<S>,
+    ) -> Result<Vec<P>, Ed25519SignatureError>
+    where
+        P: Verifier<S>,
+    {
+        // If the signature contains less than the threshold number of signers, there's
+        // no point in trying.
+        if multi_sig.signatures.len() < self.threshold as usize {
+            return Err(Ed25519SignatureError::new());
+        }
+
+        let mut matched_signers = Vec::new();
+        let mut potential_signers = self.signers.clone();
+        for signature in multi_sig.signatures.iter() {
+            let matched_signer = potential_signers.iter().find_map(|signer| {
```

Comment:
It would be a little more efficient if each sig in the multisig indicates what public key it corresponds to, so that we can avoid trying every combination. But I'm not sure it's worth changing, it's probably fine the way it is.

Reply:
Its possible, but that makes the data structure a bit uglier and bigger. I agree that this is not efficient, but it doesn't seem problematic to me since we could cap the number of signatures at the RPC entrypoint.